ipa-core/src/protocol/ipa_prf/quicksort.rs

```diff
@@ -54,65 +60,93 @@ where
     for<'a> &'a AdditiveShare<K>: IntoIterator<Item = AdditiveShare<K::Element>>,
     K: SharedValue + Field + CustomArray<Element = Boolean>,
 {
-    // expected amount of recursion: compute ceil(log_2(list.len()))
-    let expected_depth = std::mem::size_of::<usize>() * 8 - list.len().leading_zeros() as usize;
-    // create stack
-    let mut stack: Vec<(C, usize, usize)> = Vec::with_capacity(expected_depth);
-
-    // initialize stack
-    stack.push((ctx, 0usize, list.len()));
-
-    // iterate through quicksort recursions
-    while let Some((ctx, b_l, b_r)) = stack.pop() {
-        // start of quicksort function
-        // check whether sort is needed
-        if b_l + 1 < b_r {
-            // set up iterator
-            let mut iterator = list[b_l..b_r].iter().map(get_key);
-            // first element is pivot, apply key extraction function f
-            let pivot = iterator.next().unwrap();
-            // create pointer to context for moving into closure
-            let pctx = &(ctx.set_total_records(b_r - (b_l + 1)));
-            // precompute comparison against pivot and reveal result in parallel
-            let comp: BitVec<usize, Lsb0> = seq_join(
-                ctx.active_work(),
-                stream_iter(iterator.enumerate().map(|(n, k)| async move {
-                    // Compare the current element against pivot and reveal the result.
-                    let comparison =
-                        compare_gt(pctx.narrow(&Step::Compare), RecordId::from(n), k, pivot)
-                            .await?
-                            .reveal(pctx.narrow(&Step::Reveal), RecordId::from(n))
-                            .await?;
-
-                    // reveal outcome of comparison
-                    // desc = true will flip the order of the sort
-                    Ok::<_, Error>(Boolean::from(false ^ desc) == comparison)
-                })),
-            )
-            .try_collect()
-            .await?;
+    let mut ranges_for_next_pass = Vec::with_capacity(ranges_to_sort.len() * 2);
+    let mut quicksort_pass = 1;
+
+    // iterate through all of the potentially incorrectly ordered ranges
+    // make one pass, comparing each element to the pivot and splitting into two more
+    // potentially incorrectly ordered ranges
+    while !ranges_to_sort.is_empty() {
+        // compute the total number of comparisons that will be needed
+        let mut num_comparisons_needed = 0;
+        for range in &ranges_to_sort {
+            if range.len() > 1 {
```

Comment:
Is there *ever* a case where a range is zero length?  Because if you assume that ranges are all >= 1, you could do `ranges.map(|x| x.len()).sum() - ranges.len()`.

Reply:
That would be invalid input. We can assume no zero length ranges. For callers of this API, ranges correspond to sequential records with the same PRF of the match key. So even if there is just one row with a given PRF at index i- the range to describe this is (i..i+1)

Reply:
The computation you shared there is the number of comparisons in the first pass. For all subsequent passes, the ranges are no longer densely packed, and the count is different.

Reply:
That's a great suggestion, I'll make that change.